### ISSUE TITLE
current state of affairs

### ISSUE THREAD

**Issue by TheMaverickProgrammer** [TheMaverickProgrammer/battlenetwork#6]
# about
This issue will remind me where I left off and why

# where we left off
1. **[POSTPONED]** Just wrote the Chip Summon system but it is noodely, uses lambdas, and will crash depending on what item gets deleted first. It was quick prototype and I now have the knowledge to rewrite it better.

2. **[DONE]** Began rewriting the scenes into a reusable class. Originally they were quickly thrown together from the very early iteration of the engine; using a single entry function and a while-loop storing all game logic in each. In the middle of refactoring, I realized the frustration of copying the segues (screen transitions) in each one as well. Realizing that scenes needed to be composed of two parts: Activities and Segues. This inspired me to write a prototype which eventually became another pet project: [Swoosh](https://github.com/TheMaverickProgrammer/Swoosh). Now that Swoosh is stable and feature-rich, I will re-write the engine to use Swoosh for activity management and segue effects.

3. **[DONE]** Shader support on Android. ~~A big demand is support for android which required forking SFML because the committee expects to support GL ES 2.0 by 2023 the earliest. This is not soon enough. I have a working build of SFML 2.5 and the BattleNetwork engine on Android 8.0. GL ES 2.0 is not compatible with my shaders so I must rewrite them and have my copy of SFML support the new pipeline. The loading screen is there and the game runs. Other android events need to be supported like when the activity sleeps or awakens.~~. Upgrade SFML to GL ES 3.0 just enough to run the game.

# picking back up
With many other things going on and coming to an end, I'm taking a break. I'll be back on this project by March 2019 guaranteed. By then the first thing I'll do is ~~have SFML supporting Android and GL ES 2.0 fully. Then I'll have my custom shader effects working.~~

~~After mobile I'll rewrite the scenes on Desktop using Swoosh. I'll merge the two projects so that the engine is fully up-to-date and compatible on both platforms.~~

Finally, I'll rewrite the chip summons code

# plans for 2019
The engine was near-complete this year. It is missing ~~game-accurate ice tiles (partially implemented)~~, ~~tiles doing damage~~, ~~elemental modifiers for attacks~~, ~~throwable spell logic~~, and ~~chip folder screen~~. 

Everything else is considered content. Content involves registering new mobs with new animations and writing the AI. I'll release a demo with a bit more content like chips, enemies, and a few bosses at this stage.

Later in the year I'll add scripting. This will make content easier to produce.

# plans for networking
PvP and matchmaking goals depends entirely on how quickly the rest of the deadlines are reached. This feature is the lowest priority.

**Comment by TheMaverickProgrammer** [comment]
As of 12/29/2018 Swoosh is now fully integrated into the engine. It looks pretty amazing actually. [Click here to watch](https://streamable.com/4orgc)

**Comment by TheMaverickProgrammer** [comment]
Little update here for my github viewers.

After much research and consultation getting SFML to GL ES 2.0 is too much of a time sink. ES 2.0 simply lacks a lot of API features to make the migration from GL ES 1.1 easy and a lot of SFML internals would have to be rewritten.

Luckily GL ES 3.0 is more robust and offers more that makes the port easier. I've been making some headway with that port so shaders will work on Android.

Android build itself is working minus the graphical bugs mostly due to shaders missing. Watch [here](https://streamable.com/t6cm8). ⚠️ loud audio

The engine has also undergone a lot of changes thanks to contributors all over the internet whos names will be released on the readme in the next public update. Namely the battle system has a fixed pipeline that accurately depicts the expected behavior from the games. 

The engine is documented now and new content has been written as well.

**Comment by TheMaverickProgrammer** [comment]
I proudly announce the android port is finally complete. There are some things that must be considered in the long run but for now, it is playable.

Watch [here](https://streamable.com/y8989).

**Comment by TheMaverickProgrammer** [comment]
Hello watchers. Big updates on the way including the engine's very own dedicated website! 

![landing page](https://i.ibb.co/yRfnQvj/Untitled.png)

With scripting planned I expect there to be a need for community scripts and collabs. This site will be where anyone can pop in and get started making custom chips, navis, and enemies!

Later the next update will be available. This includes a robust chip summons system, new chip pipeline, more accurate state resolution (e.g. stun/flinch/drag) during battles. 

Android code is also merged together with the main code. There will an android studio project included.

Finally, all the code is fully documented and has its own documentation site for advanced programmers. [Check it out](https://battlenetwork.readthedocs.io/).

**Comment by TheMaverickProgrammer** [comment]
A few remaining items will have to wait until Dec 2019 when I pick up this project again. Here in about 2 weeks I'll merge master with all the work I've done since March and make the site go live. 

The remaining items were:
* better chip summon system -> I want to use time-based action list so that linking action sequences together has a predictable accumulative length. This will prevent poorly written summon sequences to hang the game (by never completing). The whole chip summons API itself is hackish and needs to be redesigned too.

* New chip API. Chips should be able to digest the chips before and after them (for cases like Atk+10 or Wood chips that absorb other wood chips). I have a hackish one in place that simply checks for incoming names and is hard-coded. Chips need an API anyway if scripting is going to happen.

* Scripting. Didn't touch much lua as much as I wanted to. Need to add support for this.

* Refactored BattleScene. It is a huge mess. It began very simple and then I ended up using boolean flags for basic state management. Over time I learned there were more intermediate steps in the battle sequences and as such ended with too many booleans and timers. Some logic code is in the Draw() step because of the first design before we even use Swoosh Activity's. 

Regardless of these task items, the features are there (just not coded well!) and will be present in the merge and you'll see all the stuff I've been talking on here!
